Should the typed querySelector() return mixed HTML and SVG results?

Since #11526 and #11990, if you do `document.querySelector('a')` you get back a value with the type `HTMLAnchorElement | null`, which is usually what you'd expect and want. However we've gotten burned (at least) once before on this where we're also selecting SVG elements in a page but not handling them in the code that follows.

In a page with anchors in SVG,  `document.querySelector('a')`  _should_ be returning a value of type `HTMLAnchorElement | SVGAElement | null`, and the difference in property/attribute behavior in HTML vs SVG can cause bugs. The actual issue we ran into was #6481, fixed in #6483, and then eventually became the [explicit handling of  `HTMLAnchorElement` vs `SVGAElement` in `anchor-elements`](https://github.com/GoogleChrome/lighthouse/blob/36e666c1160dab60c37ac7da028d034caafdcd2b/lighthouse-core/gather/gatherers/anchor-elements.js#L41-L72).

The reason for the current types is `typed-query-selector` looks first for a valid `HTMLElement` matching the found element identifier, if nothing is found it falls back to see if there are any matching `SVGElement`s, and only then falls back to the generic `Element` (one of the shared base classes of both `HTMLElement` and `SVGElement`):

```ts
export type TagNameToElement<Tag extends string, Fallback extends Element = Element> =
  Tag extends keyof HTMLElementTagNameMap ? HTMLElementTagNameMap[Tag] :
  Tag extends keyof SVGElementTagNameMap ? SVGElementTagNameMap[Tag] :
  Fallback
```

https://github.com/g-plane/typed-query-selector/blob/823e4b850280dab7192cd0468b8e97c196e9e8ec/parser.d.ts#L121-L125

The tags shared between `HTMLElement`s and `SVGElement`s are `<a>`, `<script>`, `<style>`, and `<title>`, and it does seem like they could run into similar issues in various gatherers that we aren't accounting for and it would be nice if the type system forced us to account for them. So we could change the type (and/or get the upstream changed) so it returns the union of found `HTMLElement`s and `SVGElement`s. Something like:

```ts
type TypeOrFallback<T, Fallback> = [T] extends [never] ? Fallback : T;

type TagNameToElement<Tag extends string> = TypeOrFallback<
  (Tag extends keyof HTMLElementTagNameMap ? HTMLElementTagNameMap[Tag] : never) |
  (Tag extends keyof SVGElementTagNameMap ? SVGElementTagNameMap[Tag] : never),
  Element
>;
```

(seems like there has to be a better pattern for that fallback, but that's the best I can come up with right now :)

Buuuut, the downside is having to deal with that in situations where we _know_ there aren't e.g. SVG anchors, like in a template we wrote in the report. For example:

https://github.com/GoogleChrome/lighthouse/blob/52b12cf79ce2cd53b0f8f8b81796691e570bb532/lighthouse-core/report/html/renderer/category-renderer.js#L328-L329

this would error, because `href` is a [readonly property](https://svgwg.org/svg2-draft/types.html#InterfaceSVGURIReference) on `SVGAElement` (you have to set the attribute). There aren't that many errors, but it does seem like a somewhat annoying situation to put on ourselves.


<details>
<summary>Current 9 errors when trying this on master:</summary>

```sh
$ tsc -p . && tsc -p lighthouse-viewer/ && tsc -p lighthouse-treemap/
lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js:91:36 - error TS2339: Property 'src' does not exist on type 'HTMLOrSVGScriptElement'.
  Property 'src' does not exist on type 'SVGScriptElement'.

91           !/^data:/.test(scriptTag.src) &&
                                      ~~~

lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js:92:36 - error TS2339: Property 'src' does not exist on type 'HTMLOrSVGScriptElement'.
  Property 'src' does not exist on type 'SVGScriptElement'.

92           !/^blob:/.test(scriptTag.src) &&
                                      ~~~

lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js:99:20 - error TS2339: Property 'src' does not exist on type 'HTMLOrSVGScriptElement'.
  Property 'src' does not exist on type 'SVGScriptElement'.

99           url: tag.src,
                      ~~~

lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js:100:20 - error TS2339: Property 'src' does not exist on type 'HTMLOrSVGScriptElement'.
  Property 'src' does not exist on type 'SVGScriptElement'.

100           src: tag.src,
                       ~~~

lighthouse-core/report/html/renderer/category-renderer.js:329:13 - error TS2540: Cannot assign to 'href' because it is a read-only property.

329     wrapper.href = `#${category.id}`;
                ~~~~

lighthouse-core/report/html/renderer/pwa-category-renderer.js:62:13 - error TS2540: Cannot assign to 'href' because it is a read-only property.

62     wrapper.href = `#${category.id}`;
               ~~~~

lighthouse-core/report/html/renderer/pwa-category-renderer.js:81:13 - error TS2339: Property 'title' does not exist on type 'HTMLAnchorElement | SVGAElement'.
  Property 'title' does not exist on type 'SVGAElement'.

81     wrapper.title = this._getGaugeTooltip(category.auditRefs, groupDefinitions);
               ~~~~~

lighthouse-core/report/html/renderer/report-renderer.js:74:17 - error TS2540: Cannot assign to 'href' because it is a read-only property.

74     metadataUrl.href = metadataUrl.textContent = report.finalUrl;
                   ~~~~

lighthouse-core/report/html/renderer/report-renderer.js:75:17 - error TS2339: Property 'title' does not exist on type 'HTMLAnchorElement | SVGAElement'.
  Property 'title' does not exist on type 'SVGAElement'.

75     metadataUrl.title = report.finalUrl;
                   ~~~~~


Found 9 errors.
```
</details>

--- a/lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js
+++ b/lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js
@@ -20,7 +20,7 @@
 
 const Gatherer = require('../gatherer.js');
 
-/* global document, window, HTMLLinkElement */
+/* global document, window, HTMLLinkElement, SVGScriptElement */
 
 /** @typedef {{href: string, media: string, msSinceHTMLEnd: number, matches: boolean}} MediaChange */
 /** @typedef {{tagName: 'LINK', url: string, href: string, rel: string, media: string, disabled: boolean, mediaChanges: Array<MediaChange>}} LinkTag */
@@ -84,7 +84,11 @@ async function collectTagsThatBlockFirstPaint() {
 
     /** @type {Array<ScriptTag>} */
     const scriptTags = [...document.querySelectorAll('head script[src]')]
-      .filter(scriptTag => {
+      .filter(/** @return {scriptTag is HTMLScriptElement} */ scriptTag => {
+        // SVGScriptElement can't appear in <head> (it'll be kicked to <body>), but keep tsc happy.
+        // https://html.spec.whatwg.org/multipage/semantics.html#the-head-element
+        if (scriptTag instanceof SVGScriptElement) return false;
+
         return (
           !scriptTag.hasAttribute('async') &&
           !scriptTag.hasAttribute('defer') &&

--- a/lighthouse-core/report/html/renderer/dom.js
+++ b/lighthouse-core/report/html/renderer/dom.js
@@ -19,6 +19,7 @@
 /* globals URL self Util */
 
 /** @typedef {HTMLElementTagNameMap & {[id: string]: HTMLElement}} HTMLElementByTagName */
+/** @template {string} T @typedef {import('typed-query-selector/parser').ParseSelector<T, Element>} ParseSelector */
 
 class DOM {
   /**
@@ -216,13 +217,18 @@ class DOM {
    * @template {string} T
    * @param {T} query
    * @param {ParentNode} context
+   * @return {ParseSelector<T>}
    */
   find(query, context) {
     const result = context.querySelector(query);
     if (result === null) {
       throw new Error(`query ${query} not found`);
     }
-    return result;
+
+    // Because we control the report layout and templates, use the simpler
+    // `typed-query-selector` types that don't require differentiating between
+    // e.g. HTMLAnchorElement and SVGAElement. See https://github.com/GoogleChrome/lighthouse/issues/12011.
+    return /** @type {ParseSelector<T>} */ (result);
   }
 
   /**

--- a/lighthouse-core/report/html/renderer/dom.js
+++ b/lighthouse-core/report/html/renderer/dom.js
@@ -227,7 +227,7 @@ class DOM {
 
     // Because we control the report layout and templates, use the simpler
     // `typed-query-selector` types that don't require differentiating between
-    // e.g. HTMLAnchorElement and SVGAElement. See https://github.com/GoogleChrome/lighthouse/issues/12011.
+    // e.g. HTMLAnchorElement and SVGAElement. See https://github.com/GoogleChrome/lighthouse/issues/12011
     return /** @type {ParseSelector<T>} */ (result);
   }
 

--- a/lighthouse-treemap/app/src/util.js
+++ b/lighthouse-treemap/app/src/util.js
@@ -8,6 +8,7 @@
 /* eslint-env browser */
 
 /** @typedef {HTMLElementTagNameMap & {[id: string]: HTMLElement}} HTMLElementByTagName */
+/** @template {string} T @typedef {import('typed-query-selector/parser').ParseSelector<T, Element>} ParseSelector */
 
 const KiB = 1024;
 const MiB = KiB * KiB;
@@ -108,13 +109,17 @@ class TreemapUtil {
    * @template {string} T
    * @param {T} query
    * @param {ParentNode=} context
+   * @return {ParseSelector<T>}
    */
   static find(query, context = document) {
     const result = context.querySelector(query);
     if (result === null) {
       throw new Error(`query ${query} not found`);
     }
-    return result;
+    // Because we control the treemap layout and templates, use the simpler
+    // `typed-query-selector` types that don't require differentiating between
+    // e.g. HTMLAnchorElement and SVGAElement. See https://github.com/GoogleChrome/lighthouse/issues/12011.
+    return /** @type {ParseSelector<T>} */ (result);
   }
 
   /**

--- a/lighthouse-treemap/app/src/util.js
+++ b/lighthouse-treemap/app/src/util.js
@@ -118,7 +118,7 @@ class TreemapUtil {
     }
     // Because we control the treemap layout and templates, use the simpler
     // `typed-query-selector` types that don't require differentiating between
-    // e.g. HTMLAnchorElement and SVGAElement. See https://github.com/GoogleChrome/lighthouse/issues/12011.
+    // e.g. HTMLAnchorElement and SVGAElement. See https://github.com/GoogleChrome/lighthouse/issues/12011
     return /** @type {ParseSelector<T>} */ (result);
   }
 

--- a/types/externs.d.ts
+++ b/types/externs.d.ts
@@ -6,7 +6,25 @@
 
 import _Crdp from 'devtools-protocol/types/protocol';
 import _CrdpMappings from 'devtools-protocol/types/protocol-mapping';
-import {ParseSelector} from 'typed-query-selector/parser';
+import {ParseSelectorToTagNames} from 'typed-query-selector/parser';
+
+/** Merge properties of the types in union `T`. Where properties overlap, property types becomes the union of the two (or more) possible types. */
+type MergeTypes<T> = {
+  [K in (T extends unknown ? keyof T : never)]: T extends Record<K, infer U> ? U : never;
+};
+
+// Helper types for strict querySelector/querySelectorAll that includes the overlap
+// between HTML and SVG node names (<a>, <script>, etc).
+// see https://github.com/GoogleChrome/lighthouse/issues/12011
+type HtmlAndSvgElementTagNameMap = MergeTypes<HTMLElementTagNameMap|SVGElementTagNameMap> & {
+  // Fall back to Element (base of HTMLElement and SVGElement) if no specific tag name matches.
+  [id: string]: Element;
+};
+type QuerySelectorParse<I extends string> = ParseSelectorToTagNames<I> extends infer TagNames ?
+  TagNames extends Array<string> ?
+    HtmlAndSvgElementTagNameMap[TagNames[number]] :
+    Element: // Fall back for queries typed-query-selector fails to parse, e.g. `'[alt], [aria-label]'`.
+  never;
 
 declare global {
   // Augment Intl to include
@@ -394,7 +412,7 @@ declare global {
 
   // Stricter querySelector/querySelectorAll using typed-query-selector.
   interface ParentNode {
-    querySelector<S extends string>(selector: S): ParseSelector<S> | null;
-    querySelectorAll<S extends string>(selector: S): NodeListOf<ParseSelector<S>>;
+    querySelector<S extends string>(selector: S): QuerySelectorParse<S>;
+    querySelectorAll<S extends string>(selector: S): NodeListOf<QuerySelectorParse<S>>;
   }
 }

